### PR TITLE
ollama: 0.1.29 -> 0.1.30

### PR DESCRIPTION
**Pull request by abysssol**
## Description of changes

Updated ollama to 0.1.30, which required updating patches. Rocm appears to work correctly, but I haven't tested cuda.

Also moved `cmake-include.patch` from ollama's build script to the patch phase of `buildGoModule`, which means that it's applied earlier. This used to cause compilation to fail, but appears to work as expected now.

### Release [0.1.30](https://github.com/ollama/ollama/releases/tag/v0.1.30) -- What's changed:

- Fixed various issues with `ollama run` on Windows
  - History now will work when pressing up and down arrow keys
  - Right and left arrow keys will now move the cursor appropriately
  - Pasting multi-line strings will now work on Windows
- Fixed issue where mounting or sharing files between Linux and Windows (e.g. via WSL or Docker) would cause errors due to having `:` in the filename.
- Improved support for AMD MI300 and MI300X Accelerators
- Improved cleanup of temporary files resulting in better space utilization

#### Important change

For filesystem compatibility, Ollama has changed model data filenames to use `-` instead of `:`. This change will be applied automatically. If downgrading to 0.1.29 or lower from 0.1.30 (on Linux or macOS only) run:

``` sh
find ~/.ollama/models/blobs -type f -exec bash -c 'mv "$0" "${0//-/:}"' {} \;
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/misc/ollama/default.nix [hunk 1]
```diff
@@ -133,31 +133,24 @@ goBuild ((lib.optionalAttrs enableRocm {
     metalFrameworks;
 
   patches = [
-    # remove uses of `git` in the `go generate` script
+    # disable uses of `git` in the `go generate` script, since nix removes the git directory
     # instead use `patch` where necessary
     ./remove-git.patch
-    # replace a hardcoded use of `g++` with `$CXX`
-    ./replace-gcc.patch
+    ./cmake-include.patch
```
Comment (drupol) [inline pkgs/tools/misc/ollama/default.nix:139]: Each patch in nixpkgs applied to the upstream source should be documented. Out-of-tree patches, fetched using `fetchpatch`, are preferred.

In each patch comment, please explain the purpose of the patch and link to the relevant upstream issue if possible. If the patch has been merged upstream but is not yet part of the released version, please note the version number or date in the comment such that a future maintainer updating the nix expression will know whether the patch has been incorporated upstream and can thus be removed from nixpkgs.

Furthermore, please use a _stable_ URL for the patch. Rather than, for example, linking to a GitHub pull request of the form `https://github.com/owner/repo/pull/pr_number.patch`, which would change every time a commit is added or the PR is force-pushed, link to a specific commit patch in the form `https://github.com/owner/repo/commit/sha.patch`.

Here are two good examples of patch comments:

```nix
mkDerivation {
  …

  patches = [
    # Ensure RStudio compiles against R 4.0.0.
    # Should be removed next 1.2.X RStudio update or possibly 1.3.X.
    (fetchpatch {
      url = "https://github.com/rstudio/rstudio/commit/3fb2397c2f208bb8ace0bbaf269481ccb96b5b20.patch";
      sha256 = "0qpgjy6aash0fc0xbns42cwpj3nsw49nkbzwyq8az01xwg81g0f3";
    })
  ];
}
```

```nix
mkDerivation {
  …

  patches = [
    # Allow building with bison 3.7
    # PR at https://github.com/GoldenCheetah/GoldenCheetah/pull/3590
    (fetchpatch {
      url = "https://github.com/GoldenCheetah/GoldenCheetah/commit/e1f42f8b3340eb4695ad73be764332e75b7bce90.patch";
      sha256 = "1h0y9vfji5jngqcpzxna5nnawxs77i1lrj44w8a72j0ah0sznivb";
    })
  ];
}
```

File: pkgs/tools/misc/ollama/default.nix [hunk 2]
```diff
@@ -191,9 +184,9 @@ goBuild ((lib.optionalAttrs enableRocm {
   };
 
   meta = {
-    changelog = "https://github.com/ollama/ollama/releases/tag/v${version}";
```
Comment (drupol) [inline pkgs/tools/misc/ollama/default.nix:194]: What's the reason of this change?

**Review by drupol — CHANGES_REQUESTED** [review]
Hey hi!

Thanks for the contribution.

It mostly LGTM, excepts the 2 nits. Let me know if you need more info, or ping me on Matrix.

File: pkgs/tools/misc/ollama/default.nix (hunk 2 above)
Comment (abysssol) [reply pkgs/tools/misc/ollama/default.nix:194]: No good reason, I just moved `changelog` down because I prefer `description` to be at the top of `meta`.

File: pkgs/tools/misc/ollama/default.nix (hunk 1 above)
Comment (abysssol) [reply pkgs/tools/misc/ollama/default.nix]: I've realized that if I changed where `disable-git.patch` inserts the `return`, `cmake-include.patch` can be removed entirely.

`disable-git.patch` will probably be necessary for as long as ollama's build script (`go generate ./...`) is used, as opposed to using a completely custom build process in nix.

Everything using `preparePatch` should just be kept in sync with the contents of `ollama/llm/patches/`, to replicate the behavior disabled by `disable-git.patch`, but without relying on `git`.

Thank you for your thorough feedback, I do appreciate it! I'm not sure if my changes were sufficient, so any further suggestions are also helpful.

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thank you @abysssol !
